packages/ra-core/src/dataProvider/useGetManyAggregate.ts

```diff
@@ -0,0 +1,186 @@
+import { UseGetManyHookValue } from './useGetMany';
+import {
+    Identifier,
+    Record,
+    ReduxState,
+    DataProviderProxy,
+    GetManyParams,
+} from '../types';
+import { useSafeSetState } from '../util/hooks';
+import { useEffect } from 'react';
+import { Refetch } from './useQueryWithStore';
+import { useDataProvider } from '.';
+
+type Callback = (args?: any) => void;
+
+interface Query {
+    ids: Identifier[];
+    resolve: Callback;
+    reject: Callback;
+}
+interface QueriesToCall {
+    [resource: string]: Query[];
+}
+
+let queriesToCall: QueriesToCall = {};
+
+/**
+ * Call the dataProvider.getMany() method and return the resolved result
+ * as well as the loading state.
+ *
+ * The return value updates according to the request state:
+ *
+ * - start: { loading: true, loaded: false, refetch }
+ * - success: { data: [data from response], loading: false, loaded: true, refetch }
+ * - error: { error: [error from response], loading: false, loaded: false, refetch }
+ *
+ * This hook will return the cached result when called a second time
+ * with the same parameters, until the response arrives.
+ *
+ * This hook aggregates and deduplicates calls to the same resource, so for instance, if an app calls:
+ *
+ * useGetMany('tags', [1, 2, 3]);
+ * useGetMany('tags', [3, 4]);
+ *
+ * during the same tick, the hook will only call the dataProvider once with the following parameters:
+ *
+ * dataProvider(GET_MANY, 'tags', [1, 2, 3, 4])
+ *
+ * @param resource The resource name, e.g. 'posts'
+ * @param ids The resource identifiers, e.g. [123, 456, 789]
+ * @param {Object} options Options object to pass to the dataProvider.
+ * @param {boolean} options.enabled Flag to conditionally run the query. If it's false, the query will not run
+ * @param {Function} options.onSuccess Side effect function to be executed upon success, e.g. { onSuccess: { refresh: true } }
+ * @param {Function} options.onError Side effect function to be executed upon failure, e.g. { onError: error => notify(error.message) }
+ *
+ * @returns The current request state. Destructure as { data, error, loading, loaded, refetch }.
+ *
+ * @example
+ *
+ * import { useGetMany } from 'react-admin';
+ *
+ * const PostTags = ({ record }) => {
+ *     const { data, loading, error } = useGetMany('tags', record.tagIds);
+ *     if (loading) { return <Loading />; }
+ *     if (error) { return <p>ERROR</p>; }
+ *     return (
+ *          <ul>
+ *              {data.map(tag => (
+ *                  <li key={tag.id}>{tag.name}</li>
+ *              ))}
+ *          </ul>
+ *      );
+ * };
+ */
+export const useGetManyAggregate = <RecordType extends Record = Record>(
+    resource: string,
+    params: GetManyParams,
+    options: UseQueryOptions<RecordType[], Error>
+): UseGetManyHookValue<RecordType> => {
+    const dataProvider = useDataProvider();
+    const queryClient = useQueryClient();
+    const { ids } = params;
+
+    return useQuery<RecordType[], Error, RecordType[]>(
```

Comment:
this approach is actually more powerful than the previous one: not only do we no longer have to mess with useState, we don't need to batch the side effects anymore either. 

Bonus: both the individual and the aggregated queries are in the cache, so after visiting a post list and displaying a post, its tags will display right away.